### PR TITLE
fix(STONEINTG-1086): annotate build PLR when failing to create snapshot

### PR DESCRIPTION
**Pull request by hongweiliu17**
Signed-off-by: Hongwei Liu <hongliu@redhat.com>

## Maintainers will complete the following section

- [ ] Commit messages are descriptive enough ([hints](https://www.freecodecamp.org/news/how-to-write-better-git-commit-messages/))
- [ ] Code coverage from testing does not decrease and new code is covered ([check the PR coverage on codecov](https://app.codecov.io/gh/redhat-appstudio/integration-service/pulls))
- [ ] [Controllers diagrams](https://github.com/konflux-ci/integration-service/tree/main/docs) are updated when PR changes controllers code  (if applicable)

### CODE REVIEW

File: internal/controller/buildpipeline/buildpipeline_adapter.go [hunk 1]
```diff
@@ -335,22 +335,20 @@ func (a *Adapter) annotateBuildPipelineRunWithSnapshot(snapshot *applicationapiv
...
 		}
 
-		if a.pipelineRun.GetDeletionTimestamp() != nil {
-			err = tekton.AnnotateBuildPipelineRunWithCreateSnapshotAnnotation(a.context, a.pipelineRun, a.client, err)
-			if err != nil {
-				return err
-			}
-			a.logger.LogAuditEvent("Updated build pipelineRun", a.pipelineRun, h.LogActionUpdate,
-				h.CreateSnapshotAnnotationName, a.pipelineRun.Annotations[h.CreateSnapshotAnnotationName])
+		err = tekton.AnnotateBuildPipelineRunWithCreateSnapshotAnnotation(a.context, a.pipelineRun, a.client, cerr)
```
Comment (dirgim) [inline internal/controller/buildpipeline/buildpipeline_adapter.go:346]: Will annotating a `pipelineRun` that's being deleted work? Afaik, after the `deletionTimestamp` is set on a resource, only updates to the finalizers field are allowed.

Comment (hongweiliu17) [reply internal/controller/buildpipeline/buildpipeline_adapter.go:346]: Thanks! Yes. I just think we should not have `a.pipelineRun.GetDeletionTimestamp() != nil`. So let me update it to `a.pipelineRun.GetDeletionTimestamp() == nil`

**Comment by hongweiliu17** [comment]
> Does this mean that the snapshot quota issue will be now reflected in annotations?

Yes,  at least users can see the error if they check build plr. And next they can know the failure from git pr/mr once we implement [STONEINTG-996](https://issues.redhat.com/browse/STONEINTG-996).

**Review by dirgim — APPROVED** [review]
LGTM

File: internal/controller/buildpipeline/buildpipeline_adapter.go (hunk 1 above)
Comment (dirgim) [reply internal/controller/buildpipeline/buildpipeline_adapter.go]: Good point, looks good now, thanks!

File: internal/controller/buildpipeline/buildpipeline_adapter.go [hunk 2]
```diff
@@ -335,23 +335,23 @@ func (a *Adapter) annotateBuildPipelineRunWithSnapshot(snapshot *applicationapiv
...
+		if a.pipelineRun.GetDeletionTimestamp() == nil {
+			err = tekton.AnnotateBuildPipelineRunWithCreateSnapshotAnnotation(a.context, a.pipelineRun, a.client, cerr)
 			if err != nil {
 				return err
 			}
 			a.logger.LogAuditEvent("Updated build pipelineRun", a.pipelineRun, h.LogActionUpdate,
 				h.CreateSnapshotAnnotationName, a.pipelineRun.Annotations[h.CreateSnapshotAnnotationName])
 		}
-
+		a.logger.Info("Cannot annotate build pipelinerun annotaion with the snapshot creation status since it has been marked as deleted")
```
Comment (dheerajodha) [inline internal/controller/buildpipeline/buildpipeline_adapter.go:354]: shouldn't this be in an `else` block?

File: internal/controller/buildpipeline/buildpipeline_adapter.go [hunk 3]
```diff
@@ -335,23 +335,23 @@ func (a *Adapter) annotateBuildPipelineRunWithSnapshot(snapshot *applicationapiv
...
 	return retry.RetryOnConflict(retry.DefaultRetry, func() error {
 		var err error
 		a.pipelineRun, err = a.loader.GetPipelineRun(a.context, a.client, a.pipelineRun.Name, a.pipelineRun.Namespace)
 		if err != nil {
 			return err
 		}
 
-		if a.pipelineRun.GetDeletionTimestamp() != nil {
-			err = tekton.AnnotateBuildPipelineRunWithCreateSnapshotAnnotation(a.context, a.pipelineRun, a.client, err)
+		if a.pipelineRun.GetDeletionTimestamp() == nil {
```
[the remark this answers is not in the document]
Comment (hongweiliu17) [reply internal/controller/buildpipeline/buildpipeline_adapter.go:346]: I was curious too.

File: internal/controller/buildpipeline/buildpipeline_adapter.go (hunk 2 above)
Comment (hongweiliu17) [reply internal/controller/buildpipeline/buildpipeline_adapter.go]: Updated. Thanks.

[the remark this answers is not in the document]
Comment (hongweiliu17) [reply internal/controller/buildpipeline/buildpipeline_adapter.go]: Updated. Thanks.

Comment (dheerajodha) [reply internal/controller/buildpipeline/buildpipeline_adapter.go]: thanks!

**Review by dheerajodha — APPROVED** [review]

**Review by Josh-Everett — APPROVED** [review]

**Pull request merged**